### PR TITLE
Create alerts two point ohhhhh yea!

### PR DESCRIPTION
**Pull request by tgsoverly**
Well this is a whopper.

This PR does more than create Alerts 2.0, it also creates two more core concepts:
1. A `ContextString` a reference to some other object or property
2. An `EvalString` a logical statement that evaluates to true or false.

These two concepts allow for the tests and alerts to reference and be triggered by many other objects in the project DAG.

This PR also has some larger organization in the project and dag constructor classes.

### CODE REVIEW

File: visivo/models/base/eval_string.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+    def get_context_strings(self) -> List[ContextString]:
+        return list(
+            map(
+                lambda m: ContextString(m),
+                re.findall(INLINE_CONTEXT_STRING_REGEX, self.value),
+            )
+        )
+
+    def evaluate(self) -> Any:
```
Comment (jpjsonic) [inline visivo/models/base/eval_string.py:41]: Is the idea here to make it safer than just running `eval()` around the string?

File: visivo/models/base/parent_model.py [hunk 2]
```diff
@@ -153,19 +153,92 @@ def name_match(trace):
...
-
-        options = {}
-        pos = spring_layout(dag)
-        draw_networkx(dag, pos, **options)
-
-        ax = pyplot.gca()
-        ax.margins(0.20)
-        pyplot.axis("off")
-        pyplot.show()
+    def show_dag(self, dag=None):
```
Comment (jpjsonic) [inline visivo/models/base/parent_model.py:156]: This is cool, are you thinking of including this chart in the viewer by default?

File: visivo/models/test.py
Comment (jpjsonic) [inline visivo/models/test.py:1]: For some reason the test model is no not rendering in the docs. I don't think we need to solve that in this PR but wanted to call that out.

File: visivo/models/base/parent_model.py (hunk 2 above)
Comment (tgsoverly) [reply visivo/models/base/parent_model.py:156]: Not yet, but eventually.  If we are using plotly, lets start going that way!

File: visivo/models/base/eval_string.py (hunk 1 above)
Comment (tgsoverly) [reply visivo/models/base/eval_string.py]: Yes, it can only execute permitted operations

File: visivo/parsers/evaluator.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
+from typing import Any
+import ast
+import operator as op
+import os
+
+SUPPORTED_NUMPY_FUNCTIONS = ["sum", "all", "mean"]
```
Comment (jpjsonic) [inline visivo/parsers/evaluator.py:6]: @tgsoverly why not allow all numpy functions? Are you thinking that would be dangerous?

File: visivo/parsers/evaluator.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            raise ValueError(f"Unsupported node type: {node.__class__.__name__}")
+
+    def eval_subscripts(node):
+        nodes = build_list_of_nodes(node)
+        nodes = nodes[::-1]  # Reverse the order of nodes
+        if hasattr(nodes[0], "id") and nodes[0].id == "project":
+            value = get_object_from_node(project, 1, nodes)
+            return value
+        else:
+            raise ValueError(f"Unsupported node type: {nodes[0].id}")
```
Comment (jpjsonic) [inline visivo/parsers/evaluator.py:62]: What's a node? Is that an object in the dag of your project?

File: visivo/commands/options.py [hunk 5]
```diff
@@ -54,13 +54,6 @@ def source(function):
     return function
 
 
-def alert(function):
```
Comment (jpjsonic) [inline visivo/commands/options.py:57]: We'll have to update some of our docs that use these options.

File: visivo/models/alert.py [hunk 6]
```diff
@@ -1,156 +1,40 @@
...
-    alerts:
-      - name: email-destination #any unique name of your choosing
-        type: email
-        subject: "[ALERT] Your Visivo Tests Have Failed" #can be any message you want
-        to: someone@your_company.com
-        port: 2525 #is this port by default
-        host: your_company_email_server.com
-        username: someones_username
-        password: {% raw %}{{ env_var('EMAIL_PASSWORD')}}{% endraw %} #We'd recommend using environment variables here for security
+### Example
```
Comment (jpjsonic) [inline visivo/models/alert.py:13]: ❤️

File: visivo/models/alert.py [hunk 7]
```diff
@@ -1,156 +1,40 @@
...
+``` yaml
+alerts:
+    - name: Example Alert
+      if: >{ anyTestFailed() && env.ENVIRONMENT == "PRODUCTION" }
+      - destinations:
```
Comment (jpjsonic) [inline visivo/models/alert.py:18]: I think maybe you wanted destinations to be a key and not a list?

File: visivo/models/base/context_string.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+INLINE_REF_REGEX = r"\${\s*ref\(([a-zA-Z0-9\s'\"\-_]+?)\)[\.\d\w]*\s*}"
+INLINE_REF_PROPS_PATH_REGEX = r"\${\s*ref\([a-zA-Z0-9\s'\"\-_]+?\)([\.\d\w\]\[]*)\s*}"
+INLINE_PATH_REGEX = r"\${\s*([a-zA-Z0-9\s'\"\-_\.\[\]]+?)\s*}"
+
+
+class ContextString:
+    """
+    Represents a string that contains a reference to named object in the project model.
+    Currently you can reference another object by using the ref() function.  Example:
```
Comment (jpjsonic) [inline visivo/models/base/context_string.py:15]: Could be good to use [admonitions](https://squidfunk.github.io/mkdocs-material/reference/admonitions/) to call out the deprecation of the old `ref()` function
ex- 
``` 
!!! warning "<v1.037 behavior" 
    In previous versions of visivo you were able to reference other models using the naked `ref()` 
    function
```

File: visivo/models/base/eval_string.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+"""
+EvalString represents a string that contains expressions to be evaluated.
+
+This class is used for handling strings that may contain references or expressions
+enclosed in '>{' and '}' delimiters. It provides functionality to validate the
+string format and extract embedded ContextString instances.
+
+Available injected methods:
+    - any_test_failed() -> bool
```
Comment (jpjsonic) [inline visivo/models/base/eval_string.py:17]: I wonder if it makes sense to nest these in a `run` or `test` object that gets updated when you run visivo cli commands. Then you could reference it like `test.any_test_failed`

**Review by jpjsonic — COMMENTED** [review]
Looks like you're making a ton of progress!

File: visivo/parsers/evaluator.py (hunk 4 above)
Comment (tgsoverly) [reply visivo/parsers/evaluator.py:100]: A node in this context is an node in the Abstract Syntax Tree (AST).  This is how language gets parsed, so I am finding what can be done in our EvalStrings

File: visivo/parsers/evaluator.py (hunk 3 above)
Comment (tgsoverly) [reply visivo/parsers/evaluator.py]: I want to make sure they work, and have a test for them.  Easy to add, we can add a bunch of them.

File: test-projects/integration/project.visivo.yml [hunk 10]
```diff
@@ -163,7 +164,8 @@ dashboards:
                   order_by:
                     - query(x asc)
                   tests:
-                    - logic: assert_that(Double Simple Line.props.x[2]).is_equal_to(3)
+                    - assertions:
+                        - ">{ ${ref(Double Simple Line).props.x[2]} == 3 }"
```
Comment (jpjsonic) [inline test-projects/integration/project.visivo.yml:168]: ❤️

**Review by jpjsonic — APPROVED** [review]
🦜

File: visivo/models/base/eval_string.py (hunk 9 above)
Comment (tgsoverly) [reply visivo/models/base/eval_string.py:18]: I am not sure I followed that.

File: visivo/models/base/context_string.py (hunk 8 above)
Comment (tgsoverly) [reply visivo/models/base/context_string.py:16]: I like this.

**Pull request merged**